segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java

```diff
@@ -0,0 +1,186 @@
+     *
+     * @param tableName Name of the table segment.
+     * @param tableStore Instance of the {@link TableStore}.
+     */
+    public TableBasedMetadataStore(String tableName, TableStore tableStore) {
+        this.tableStore = Preconditions.checkNotNull(tableStore, "tableStore");
+        this.tableName = Preconditions.checkNotNull(tableName, "tableName");
+    }
+
+    /**
+     * Reads a metadata record for the given key.
+     *
+     * @param key Key for the metadata record.
+     * @return Associated {@link io.pravega.segmentstore.storage.metadata.BaseMetadataStore.TransactionData}.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    @Override
+    protected TransactionData read(String key) throws StorageMetadataException {
+        ensureInitialized();
+        List<BufferView> keys = new ArrayList<>();
+        keys.add(new ByteArraySegment(key.getBytes()));
+        try {
+            List<TableEntry> retValue = this.tableStore.get(tableName, keys, timeout).get();
+            if (retValue.size() == 1) {
+                TableEntry entry = retValue.get(0);
+                if (null != entry) {
+                    val arr = entry.getValue();
+                    TransactionData txnData = serializer.deserialize(arr);
+                    txnData.setDbObject(entry.getKey().getVersion());
+                    txnData.setPersisted(true);
+                    return txnData;
+                }
+            }
+        } catch (IllegalStateException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new StorageMetadataException("Error while reading", e);
+        }
+
+        return TransactionData.builder()
+                .key(key)
+                .persisted(true)
+                .dbObject(TableKey.NOT_EXISTS)
+                .build();
+    }
+
+    /**
+     * Writes transaction data from a given list to the metadata store.
+     *
+     * @param dataList List of transaction data to write.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    @Override
+    protected void writeAll(Collection<TransactionData> dataList) throws StorageMetadataException {
+        ensureInitialized();
+        List<TableEntry> toUpdate = new ArrayList<>();
+        HashMap<TableEntry, TransactionData> entryToTxnDataMap = new HashMap<TableEntry, TransactionData>();
+        HashMap<TableKey, TransactionData> deletedKeyToTxnDataMap = new HashMap<TableKey, TransactionData>();
+        List<TableKey> keysToDelete = new ArrayList<>();
+        try {
+            for (TransactionData txnData : dataList) {
+                Preconditions.checkState(null != txnData.getDbObject());
+
+                long version = ((Long) txnData.getDbObject()).longValue();
+                if (null == txnData.getValue()) {
+                    val toDelete = TableKey.versioned(new ByteArraySegment(txnData.getKey().getBytes()),
```

Comment:
1. encoding here 
2. Use `TableKey.Unversioned` if you want to specify `NO_VERSION`

Reply:
Or did you mean to specify a version (you have a version variable)?

Reply:
Fixed